lib/si-service/src/shutdown.rs

```diff
@@ -0,0 +1,95 @@
+//! Graceful service/server shutdown using cancellation tokens, task trackers, driven by Unix
+//! signal handling.
+
+use std::{error, future::Future, io, time::Duration};
+
+use telemetry::prelude::*;
+use thiserror::Error;
+use tokio::{
+    signal::unix::{self, SignalKind},
+    time::timeout,
+};
+use tokio_util::{sync::CancellationToken, task::TaskTracker};
+
+/// An error that can be returned when gracefully shutting down.
+///
+/// See [`graceful`] for more details.
+#[derive(Debug, Error)]
+pub enum ShutdownError {
+    /// When a signal handler fails to be correcly setup
+    #[error("failed to setup signal handler: {0}")]
+    Signal(#[source] io::Error),
+    /// When there is a telemetry-related error
+    #[error("telemetry shutdown error: {0}")]
+    Telemetry(#[source] Box<dyn error::Error + Send + Sync + 'static>),
+    /// When the timeout to wait for graceful shutdown has been exceeded
+    #[error("graceful shutdown timeout elapsed: {0:?}")]
+    TimeoutElapsed(Duration),
+}
+
+impl ShutdownError {
+    fn telemetry<E>(err: E) -> Self
+    where
+        E: error::Error + Send + Sync + 'static,
+    {
+        Self::Telemetry(Box::new(err))
+    }
+}
+
+/// Gracfully shutdown a service that may be running multiple tasks and in-flight work.
+///
+/// # Platform-specific behavior
+///
+/// This function sets up a signal handler for both `SIGINT` (i.e. `Ctrl+c`) and `SIGTERM` so usage
+/// of this function with other code intercepting these signals is *highly* discouraged.
+pub async fn graceful<Fut, E>(
```

Comment:
Love it.

Reply:
I predict this'll be used everywhere shortly